src/Nancy/Routing/RouteResultHelper.cs

```diff
@@ -0,0 +1,56 @@
+namespace Nancy.Routing
+{
+    using System;
+    using System.Collections.Concurrent;
+    using System.Reflection;
+    using System.Threading.Tasks;
+
+    /// <summary>
+    /// Helper methods to coerce route results
+    /// </summary>
+    public static class RouteResultHelper
+    {
+        private static readonly MethodInfo ConvertMethod = typeof(RouteResultHelper).GetMethod("Convert", BindingFlags.NonPublic | BindingFlags.Instance | BindingFlags.Static);
+        private static readonly ConcurrentDictionary<Type, MethodInfo> ConverterCache = new ConcurrentDictionary<Type, MethodInfo>();
+
+        /// <summary>
+        /// Coerces the result.
+        /// </summary>
+        /// <param name="routeResult">The result that should be coereced.</param>
+        /// <returns>A <see cref="Task{TResult}"/> of <see cref="object"/>.</returns>
+        public static async Task<object> CoerceResult(object routeResult)
+        {
+            if (routeResult == null)
+            {
+                return null;
+            }
+
+            var resultType = routeResult.GetType();
+
+            if (resultType != typeof(Task))
+            {
+                return routeResult;
+            }
+
+            Type innerType = resultType.GetGenericArguments()[0];
+
+            var converter = ConverterCache.GetOrAdd(innerType, t =>
+            {
+                return ConvertMethod.MakeGenericMethod(t);
+            });
+
+            return await (Task<object>)converter.Invoke(null, new [] { routeResult });

```

Comment:
Is it necessary to `await` here? Why not return the task and drop the `async` from the signature (and add `Task.FromResult` on the other `return`s)?